src/library/scala/collection/immutable/Seq.scala

```diff
@@ -34,6 +34,28 @@ trait IndexedSeq[+A] extends Seq[A]
 
   final override def toIndexedSeq: IndexedSeq[A] = this
 
+  override def canEqual(that: Any): Boolean = that match {
+    case otherIndexedSeq: IndexedSeq[_] => length == otherIndexedSeq.length && super.canEqual(that)
+    case _ => super.canEqual (that)
+  }
+
+  override def equals(o: Any): Boolean = o match {
+    case otherIndexedSeq: IndexedSeq[_] =>
+      (this eq otherIndexedSeq) || {
+        var equal = length == otherIndexedSeq.length
+        if (equal) {
+          val length = this.length
+          var index = 0
+          while (index < length && equal) {
+            equal = this (index) == otherIndexedSeq(index)
```

Comment:
For large `Vector`-s, this could well be slower than walking the iterators. `VectorIterator.next` is typically just:

```
    val res = display0(lo).asInstanceOf[A]
    lo += 1
```

Whereas `Vector.apply` needs to recalculate the path to the element and walk through the tree from root each time.

I wonder if the performance problems you witnessed would be solved _just_ by including the fast paths for `this.length != that.length`.